### PR TITLE
Test results on add test to identifier page

### PR DESCRIPTION
**Pull request by oliverviljamaa**
## Context

The _add test to identifier page_ currently only has the user creation part implemented, but it _should_ handle adding the test to the just-created user as well.

## Changes

**Adding the test has been added to the page.**

For that:
* `TestFields` have been extracted from `AddTestForm`
* `oneOf` support is added to the form creation from JSON schema

### CODE REVIEW

File: src/testHelpers.tsx [hunk 1]
```diff
@@ -17,3 +18,32 @@ export function renderWrapped(element: ReactElement): RenderResult {
     rerender: (newElement: ReactElement): void => renderResult.rerender(wrap(newElement)),
   };
 }
+
+export function aTestType(): TestType {
+  return {
+    id: 'some-test-type-id',
+    name: 'PCR test',
+    resultsSchema: {
```
Comment (oliverviljamaa) [inline src/testHelpers.tsx:26]: @StevePole Can you confirm this is the correct schema?

File: src/testHelpers.tsx [hunk 2]
```diff
@@ -17,3 +18,32 @@ export function renderWrapped(element: ReactElement): RenderResult {
...
+      title: 'Test result',
```
Comment (oliverviljamaa) [inline src/testHelpers.tsx:27]: We seem to require a `title` on the top level of `resultsSchema` as well, is it fine to keep it the same as the property one in our case?

File: src/testing/TestFields.tsx [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+  if (required) {
+    validation = validation.required();
+  }
+
+  return validation;
+}
+
+function initialPropertyValue({ type, oneOf }: FieldSchema) {
+  if (oneOf) {
+    // FIXME: What to do here?
```
Comment (oliverviljamaa) [inline src/testing/TestFields.tsx:136]: @Tankenstein @StevePole What do you think about this? Added this, as we don't want either of the options to be preselected on init.

**Comment by StevePole** [comment]
Why does the textarea have a different border to the text input, is one focused in your screenshot?

File: src/testHelpers.tsx (hunk 2 above)
Comment (StevePole) [reply src/testHelpers.tsx:27]: It's fine for an object schema to have a title.  When you say it's 'required' what do you mean, what requires it?

File: src/testHelpers.tsx (hunk 1 above)
Comment (StevePole) [reply src/testHelpers.tsx:26]: Schema looks fine.  The `type: boolean` isn't strictly required, as the sub schema inside the oneOf dictates the type.  I'm 99% sure it's valid JSON schema, but could possibly confuse other JSON schema libraries.

File: src/testing/TestFields.tsx [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+                  />
+                  {option.title}
+                </Label>
+              ))}
+              {fieldError(key)}
+            </Box>
+          );
+        }
+
+        if (value.type === 'boolean') {
```
Comment (StevePole) [inline src/testing/TestFields.tsx:45]: Yeah, by having type: "boolean" in your oneOf schema, I think you're going to render a checkbox and a radio control?

Comment (StevePole) [reply src/testing/TestFields.tsx:45]: Oh, actually no, because you return...

Comment (oliverviljamaa) [reply src/testing/TestFields.tsx:45]: Recommendations here are very welcome, as it was a bit tricky to get to work.

File: src/testing/TestFields.tsx (hunk 3 above)
Comment (StevePole) [reply src/testing/TestFields.tsx:136]: That's the correct logic, given the sensitivity of the data we don't want to preselect anything.

File: src/testing/TestFields.tsx [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+  testType: TestType;
+}
+
+export const TestFields: FC<TestFieldsProps> = ({ form, testType }) => {
+  const fieldError = (key: keyof TestType['resultsSchema']['properties']) =>
+    form.touched[key] && form.errors[key] ? <Text>{form.errors[key]}</Text> : null;
+
+  return (
+    <>
+      {Object.entries(testType.resultsSchema.properties).map(([key, value]) => {
```
Comment (StevePole) [inline src/testing/TestFields.tsx:20]: I might have created a JsonSchemaForm component, that just takes a schema, a (nullable) initial model, and fires an onChange when the model changes.

The TW one also takes a 'submitted' prop, which triggers validation on any fields that haven't already done it.

File: src/testing/AddTestToIdentifierPage.test.tsx [hunk 6]
```diff
@@ -14,77 +15,140 @@ jest.mock('../common/useConfig');
...
 
-    const { getByText, getByLabelText, queryByText, findByText } = renderWrapped(
-      <AddTestToIdentifierPage />
-    );
+    renderWrapped(<AddTestToIdentifierPage />);
 
-    const identifierInput = getByLabelText(/identification code/i) as HTMLInputElement;
-    const submit = () => fireEvent.click(getByText(/add test result/i));
+    const identifierInput = screen.getByLabelText(/identification code/i);
+    const button = screen.getByRole('button');
```
Comment (StevePole) [inline src/testing/AddTestToIdentifierPage.test.tsx:36]: I assume this is the submitButton, could you rename?  I found it unclear further down the test.

File: src/testing/AddTestToIdentifierPage.test.tsx [hunk 7]
```diff
@@ -14,77 +15,140 @@ jest.mock('../common/useConfig');
...
+
+    userEvent.type(identifierInput, '39210030814'); // valid id code
+    await waitForElementToBeRemoved(screen.queryByText(/check the identification code/i)); // no validation error
+
+    await userEvent.type(notesInput, 'Some notes');
+    expect(button).toBeDisabled(); // as a radio option has not been selected
+
+    userEvent.click(negativeOption);
+    userEvent.click(positiveOption);
+    await waitFor(() => expect(button).not.toBeDisabled());
```
Comment (StevePole) [inline src/testing/AddTestToIdentifierPage.test.tsx:55]: I was unclear here.  Firstly, why do you need to click both options?  Then I wasn't sure if button was a radio button without going up to check.

File: src/testing/AddTestToIdentifierPage.test.tsx [hunk 8]
```diff
@@ -14,77 +15,140 @@ jest.mock('../common/useConfig');
...
     expect(identifierInput).toHaveFocus();
   });
 
-  it('creates a user for the authentication method from config and the filled identifier', async () => {
-    const authenticationDetails = { method: 'ESTONIAN_ID', identifier: '39210030814' };
-    const scope = nock(/./)
-      .post('/api/v1/users', { authenticationDetails })
-      .matchHeader('Authorization', 'Bearer some-token')
-      .reply(201, { id: 'some-id', authenticationDetails });
+  it('creates a user for the authentication method from config and the filled identifier and adds a test for that user', async () => {
```
Comment (StevePole) [inline src/testing/AddTestToIdentifierPage.test.tsx:30]: Generally a bit hard to follow as the test does a lot of things.

File: src/testing/AddTestToIdentifierPage.test.tsx (hunk 7 above)
Comment (oliverviljamaa) [reply src/testing/AddTestToIdentifierPage.test.tsx:55]: This is unclear indeed — I had a bug where none of the changes after the first had any effect, so this part of the test ensures it won't happen again.

I'll rename `button` to `submitButton`, should I also rename `*Option`s to `*Radio` or `*RadioOption`?

File: src/testHelpers.tsx (hunk 2 above)
Comment (oliverviljamaa) [reply src/testHelpers.tsx:27]: Our current `ObjectSchema` definition (type of `testType.resultsSchema`). Would you make it optional instead?

**Comment by oliverviljamaa** [comment]
> Why does the textarea have a different border to the text input, is one focused in your screenshot?

Good catch, it was because the input's border colour had been manually overwritten. Unified them to the same colour that the radio button circle uses and updated the description 👍

File: src/testing/AddTestToIdentifierPage.test.tsx (hunk 8 above)
Comment (oliverviljamaa) [reply src/testing/AddTestToIdentifierPage.test.tsx:30]: Indeed — I'll try to split some of the logic out!

File: src/testing/TestFields.tsx (hunk 5 above)
Comment (oliverviljamaa) [reply src/testing/TestFields.tsx:20]: Yup, we considered this as a next refactor — will look into it, but probably postpone it till after the launch.

File: src/testing/TestFields.tsx [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+              {value.title} {testType.resultsSchema.required?.includes(key) && '*'}
+              {value.oneOf.map((option) => (
+                <Label key={`${option.const}`}>
+                  <Radio
+                    checked={form.getFieldProps(key).value === option.const}
+                    value={typeof option.const === 'number' ? option.const : `${option.const}`}
+                    onChange={({ target }) =>
+                      form.setFieldValue(
+                        key,
+                        typeof option.const === 'boolean' ? target.value === 'true' : target.value
```
Comment (StevePole) [inline src/testing/TestFields.tsx:33]: Why do you need this, rather than just `option.const` ?

File: src/testing/TestFields.tsx [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+      {Object.entries(testType.resultsSchema.properties).map(([key, value]) => {
+        if (value.oneOf) {
+          return (
+            <Box key={key}>
+              {value.title} {testType.resultsSchema.required?.includes(key) && '*'}
+              {value.oneOf.map((option) => (
+                <Label key={`${option.const}`}>
+                  <Radio
+                    checked={form.getFieldProps(key).value === option.const}
+                    value={typeof option.const === 'number' ? option.const : `${option.const}`}
```
Comment (StevePole) [inline src/testing/TestFields.tsx:29]: Same question here, does the value need to be a string?

File: src/testing/TestFields.tsx (hunk 4 above)
Comment (StevePole) [reply src/testing/TestFields.tsx:45]: The way I have this stuff implemented is something more like this (pseudocode):
```
useState(model, setModel, props.model);

const onPropertyChange = (property, value) => {
  const newModel = { ...model, [property] = value};
  setModel(newModel);
  this.onChange(newModel);
});

props.schema.properties.map((childSchema, property) => {
  <Field 
    schema={childSchema} 
    model={model[property]}
    onChange={(value) => onPropertyChange(property, value) 
  />
});
```
Where the Field component handles the control type switch, and any data type mapping issues (like radio needing a string).

File: src/testing/TestFields.tsx (hunk 10 above)
Comment (oliverviljamaa) [reply src/testing/TestFields.tsx:29]: Yup, that's the tricky part — it needs to be a number or a string, so for boolean, we need to cast it to a string and also convert back to a boolean in the change handler.

File: src/testing/TestFields.tsx (hunk 4 above)
Comment (StevePole) [reply src/testing/TestFields.tsx:45]: In the real one we also pass an isValid with the on Change

`this.onChange(newModel, isValid)`

Which can be calculated with an off the shelf JSON schema validator.  That makes it easy to do stuff like enabling/disabling submit button.

File: src/testHelpers.tsx (hunk 2 above)
Comment (StevePole) [reply src/testHelpers.tsx:27]: Yeah, it's definitely optional in the context of broader JSON schema.  There are a number of schemas that don't normally have a type: allOf, oneOf, anyOf.

Comment (oliverviljamaa) [reply src/testHelpers.tsx:27]: I'll make it optional here as well then — we're not using it for the `oneOf` case. Thanks!

Comment (oliverviljamaa) [reply src/testHelpers.tsx]: ✅ Made it optional.

File: src/testing/TestFields.tsx (hunk 3 above)
Comment (oliverviljamaa) [reply src/testing/TestFields.tsx]: ✅ Removed the `FIXME`.

File: src/testing/AddTestToIdentifierPage.test.tsx (hunk 6 above)
Comment (oliverviljamaa) [reply src/testing/AddTestToIdentifierPage.test.tsx]: ✅ Renamed!

File: src/testing/TestFields.tsx (hunk 9 above)
Comment (oliverviljamaa) [reply src/testing/TestFields.tsx:34]: https://github.com/cov-clear/web/pull/62#discussion_r421372523

File: src/testing/TestFields.tsx (hunk 4 above)
Comment (oliverviljamaa) [reply src/testing/TestFields.tsx:46]: Added a `TODO` to extract general components out.

File: src/testing/AddTestToIdentifierPage.test.tsx (hunk 7 above)
Comment (oliverviljamaa) [reply src/testing/AddTestToIdentifierPage.test.tsx]: Added a comment and it's in a shorter test now, so hopefully clear.

File: src/testing/AddTestToIdentifierPage.test.tsx (hunk 8 above)
Comment (oliverviljamaa) [reply src/testing/AddTestToIdentifierPage.test.tsx]: Split validation testing, added helper functions for readability, and moved more common setup, including the render, to `beforeEach` 👍

**Comment by oliverviljamaa** [comment]
Closed #42.
